components/browser/session/src/main/java/mozilla/components/browser/session/LegacySessionManager.kt

```diff
@@ -398,20 +399,44 @@ class LegacySessionManager(
     }
 
+    @Suppress("LongMethod") // Yes, this method is pretty long. I hope we can delete it soon.
     private fun findNearbySession(index: Int, predicate: (Session) -> Boolean): Int {
-        val maxSteps = max(values.lastIndex - index, index)
+        // Okay, this is a bit stupid and complex. This implementation intends to implement the same behavior we use in
+        // BrowserStore - which is operating on a list without custom tabs. Since LegacySessionManager uses a list with
+        // custom tabs mixed in, we need to shift the index around a bit.
+
+        // First we need to determine the number of custom tabs that are *before* the index we removed from. This is
+        // later needed to calculate the index we would have removed from in a list without custom tabs.
+        var numberOfCustomTabsBeforeRemovedIndex = 0
+        for (i in 0..min(index - 1, values.lastIndex)) {
+            if (values[i].isCustomTabSession()) {
+                numberOfCustomTabsBeforeRemovedIndex++
+            }
+        }
 
-        if (maxSteps <= 0) {
-            return NO_SELECTION
+        // Now calculate the index and list without custom tabs included.
+        val indexWithoutCustomTabs = index - numberOfCustomTabsBeforeRemovedIndex
+        val sessionsWithoutCustomTabs = sessions
+
+        // Code run before this one would have re-used the selected index if it is still a valid index that satisfies
+        // the predicate - however that list contained custom tabs. Now without custom tabs we may be able to re-use
+        // the exiting index. Let's check if we can:
+        if (indexWithoutCustomTabs >= 0 &&
+            indexWithoutCustomTabs <= sessionsWithoutCustomTabs.lastIndex &&
+            predicate(sessionsWithoutCustomTabs[indexWithoutCustomTabs])) {
+            return values.indexOf(sessionsWithoutCustomTabs[indexWithoutCustomTabs])
         }
 
-        // Try sessions oscillating near the index.
-        for (steps in 1..maxSteps) {
-            listOf(index - steps, index + steps).forEach { current ->
+        // Finally try sessions (without custom tabs) oscillating near the index.
+        val range = 1..max(sessionsWithoutCustomTabs.lastIndex - indexWithoutCustomTabs, indexWithoutCustomTabs)
+        for (steps in range) {
+            listOf(indexWithoutCustomTabs - steps, indexWithoutCustomTabs + steps).forEach { current ->
                 if (current >= 0 &&
-                    current <= values.lastIndex &&
-                    predicate(values[current])) {
-                    return current
+                    current <= sessionsWithoutCustomTabs.lastIndex &&
+                    predicate(sessionsWithoutCustomTabs[current])) {
+                    // Now that we have a match we need to re-calculate the index in the list that includes custom tabs.
+                    val session = sessionsWithoutCustomTabs[current]
+                    return values.indexOf(session)
```

Comment:
I'm a bit confused how this re-write helps. Is `predicate` necessary?

This `predicate` is only specified twice: once where it's checking we're not in a custom tab, and the other to check that we're in `private` browsing or not.

Reply:
I wonder if this code could be clearer if it were moar functional.

Reply:
The goal of the rewrite was to have the same behavior here as in `TabListReducer` of `browser-state` so that the new test (see #3567) passes.

I agree that this could be rewritten in a way more sane way. But I tried to write as little code as possible since this is the old code that is going away and I didn't want to refactor all that. So yeah, this legacy code is now a bit stupid and does definitely not optimized.

The same code in `TabListReducer` is thankfully much simpler:
https://github.com/mozilla-mobile/android-components/blob/master/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TabListReducer.kt#L171